app/src/main/java/com/example/dungeontest/graph/OpenAiRequestBuilder.kt

```diff
@@ -0,0 +1,133 @@
+        This drawing describes a network of rooms, which can be thought of as an undirected graph (each room is a node, hallways are edges). Each room is represented by box containing a text label, and each hallway is represented by 2 parallel lines.
+
+        All of your output must be in a single JSON object.
+
+        First, please create a numbered list of the rooms in the drawing (root-level json key is "rooms", each list entry should have a numeric id in addition to a name).
+        Second, for each room (A), make a list for A's top side hallways of the names and ids of the rooms on the other ends of those hallways; then make a list for A's left side hallways of the names and ids of the rooms on the other ends of those hallways; then make a list for A's right side hallways of the names and ids of the rooms on the other ends of those hallways; then make a list for A's bottom side hallways of the names and ids of the rooms on the other ends of those hallways; finally, create a single list with all of A's neighbors on those different sides. (root level json key is "adjacency_analysis").
+        Third, please make a JSON array describing that graph (root level key for the array is "final_result"), where each entry in the array is a json object containing a) a unique numeric node id (json key "id"), b) the room label from the drawing (json key "label"), and c) the id's of the rooms which it's connected to by hallways (json key "neighbors").
+    """.trimIndent()
+
+    private val client = OkHttpClient
+        .Builder()
+        .connectTimeout(40, TimeUnit.SECONDS)
+        .readTimeout(400, TimeUnit.SECONDS)
+        .build()
+    /* Probably can adjust this? */
+    private val maxTokenCount = 3000
+
+    fun sendOpenAIRequest(apiKey: String, apiVersion: String, imagePath: String, callback: (String?) -> Unit) {
+        val imageFile = File(imagePath)
+        if (!imageFile.exists()) {
+            println("Image file not found: $imagePath")
+            callback(null)
+            return
+        }
+
+        val imageBytes = imageFile.readBytes()
+        Log.v("OpenAiRequestBuilder", imageBytes.size.toString())
+        if (imageBytes.size > 20 * 1024 * 1024) {
+            println("Image file is too large: $imagePath")
+            callback(null)
+            return
+        }
+
+        val supportedFormats = listOf("png", "jpeg", "gif", "webp")
+        val fileExtension = imageFile.extension.lowercase()
+        if (fileExtension !in supportedFormats) {
+            println("Unsupported image format: $fileExtension")
+            callback(null)
+            return
+        }
+
+        val imageBase64 = Base64.getEncoder().encodeToString(imageBytes)
+        val dataUri = "data:image/$fileExtension;base64,$imageBase64"
+
+        val jsonBody = JSONObject().apply {
+            put("model", apiVersion)
+            put("response_format", JSONObject().apply {
+                put("type", "json_object")
+            })
```

Comment:
I think the previous 3 lines can be simplified to one:
`put("response_format", JSONObject().put("type", "json_object"))`  

please let me know if I'm misunderstanding the use of `apply {}` here

Reply:
This is just how I saw to initiate a JSONObject with values. I boke it into 3 lines for readability

Reply:
it feels to me like the 1 line version is at least as readable on its own, while making the code as a whole a bit easier to see holistically and scroll through.
@pepsipotty which approach is easier for you to read?

Reply:
@BareBeaverBat @EricMiddie, I personally find 
`            put("response_format", JSONObject().apply {
                put("type", "json_object")
            })
`

easier to read and more "modular". However, assuming we're only gonna add  one key value pair `put("response_format", JSONObject().put("type", "json_object"))`, then the @BareBeaverBat's recommendation is more concise and apt. Just my two cents

Reply:
I have no idea how to retain indentation in these comments. Please assume my response above has the appopriate indentation for Eric's approach